### ISSUE TITLE
0.195.0 regression: AiModelBuilder(string) reinterpreted as YAML config path (was license key); Transformer facade training degenerates to chance

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1328]
### Bug 1 — AiModelBuilder(string) constructor overload collision

In 0.195.0 a new constructor was added:

\`\`\`csharp
public AiModelBuilder(string configFilePath, AiDotNetLicenseKey? licenseKey = null)
\`\`\`

This silently shadows callers that were previously passing a license-key STRING expecting the legacy 1-arg constructor. The library now interprets the license key as a file path and throws FileNotFoundException: YAML config file not found.

Impact: every downstream consumer that was passing a license key as string (per the previous public API) now crashes immediately.

Repro:
\`\`\`csharp
var builder = new AiModelBuilder<float, Tensor<float>, Tensor<float>>(\"some-license-string\");
// throws FileNotFoundException: YAML config file not found.
\`\`\`

Workaround: wrap the string via typed AiDotNetLicenseKey to disambiguate which overload is hit.

Proper fix: either (a) require explicit configFilePath: named-arg use; (b) split into distinct method names like FromYamlConfig(path) vs AiModelBuilder(AiDotNetLicenseKey?). Overload resolution between a license-string and a path-string is a footgun.

### Bug 2 — Transformer training degenerates to chance output (PPL=V)

After working around Bug 1, Transformer<float> trained via the facade produces a degenerate output:

- top-1 = 0/V (literally never correct)
- top-5 = 100% (concentrates on a fixed 5-byte subset)
- PPL = V exactly (PPL = 256.00 with zero variance across seeds on byte-LM)
- ECE = 0.0039 (uniform-distribution calibration)
- train wall: 5-9s for 30 iterations on 100KB corpus (so SGD IS running, just not learning anything useful)

This pattern (top-5=100% concentrated on fixed bytes, top-1=0%) suggests the model outputs the same fixed distribution regardless of input. Possible causes:
- Cross-entropy gradient not flowing back through attention/embedding
- Data pipeline misaligns input/target
- SequencePoolingMode=null default isn't producing a per-sequence prediction; loss read from unintended position

This was working in 0.176.0 (see HarmonicEngine test Phase_L4_HeadToHead_DepthSweep_Screen.DepthSweep_HRE_vs_ClassicalTransformer_HeadToHead). Both Bug 1 and Bug 2 are 0.195.0 regressions.

### Why this matters

HarmonicEngine treats the AiDotNet facade as the single point of optimization. Every \"HRE vs trained classical\" head-to-head measurement is invalid if the classical arm doesn't actually train. Two facade-using test suites in HarmonicEngine produce invalid numbers because of these bugs.

### Suggested triage priority

P0. Currently blocks HRE-vs-classical research pipeline at byte-LM scale.

**Comment by ooples** [comment]
**Retracting Bug 1**: this is user error on my end. `AiDotNetLicenseKey` is the intended typed wrapper for license keys; passing a raw `string` to `AiModelBuilder(string)` was always wrong (or got lucky on the old overload set). The new overload that interprets string as a YAML config path is a reasonable design choice. The facade-side fix is purely on the HarmonicEngine consumer.

**Bug 2 stands**: facade-trained Transformer produces degenerate chance output (PPL=V exactly, top-5=100% concentrated on fixed bytes, top-1=0%, zero variance across seeds). This is the real regression. Investigating now on a parallel worktree.

**Comment by ooples** [comment]
## Diagnostic findings — confirmed regression + narrowed scope

I built diagnostic tests on branch [fix/1328-transformer-facade-training-degeneracy](https://github.com/ooples/AiDotNet/tree/fix/1328-transformer-facade-training-degeneracy) (commit c1570da0b). Two new tests in `tests/AiDotNet.Tests/IntegrationTests/NeuralNetworks/TransformerByteLMDiagnostic1328Tests.cs`:

### Test 1: Loss trajectory at default LR
Same setup as the existing `Transformer_ByteLM_DefaultPooling_TrainsBelowLnV` but logs every 50 epochs. Result:

```
epoch=000  avgNll=2.5148  ln(V)=2.0794  ratio=1.209  (init)
epoch=001  avgNll=2.5110  ratio=1.208  top-1=1/8=12.5%
epoch=050  avgNll=2.5661  ratio=1.234  top-1=1/8=12.5%
epoch=100  avgNll=2.5817  ratio=1.242  top-1=1/8=12.5%
epoch=150  avgNll=2.5907  ratio=1.246  top-1=1/8=12.5%
epoch=200  avgNll=2.5971  ratio=1.249  top-1=1/8=12.5%
epoch=250  avgNll=2.6022  ratio=1.251  top-1=1/8=12.5%
```

Loss slowly **climbs** (anti-learning). Top-1 stays at 1/8 = 12.5% **exactly** at every epoch — the model is predicting the same byte regardless of which of the 8 distinct inputs it sees.

### Test 2: High LR + disabled warmup (rules out LR throttling)
Same setup but `lr=0.1` (100× default) and `warmupSteps=1` (disabled). Result:

```
epoch=000  avgNll=2.4416  ratio=1.175  (init)
epoch=001  avgNll=10.5113  ratio=5.055  top-1=1/8=12.5%
epoch=025  avgNll=18.1329  ratio=8.720  top-1=1/8=12.5%
epoch=050+  avgNll=18.1329  (stuck saturated)
```

avgNll=18.13 corresponds to p_target ≈ exp(-18.13) ≈ 1.2e-8. Model puts essentially **all probability on a single byte that's never the target**. Top-1 still 1/8 = 12.5%.

### Diagnostic conclusion
1. **Gradients ARE flowing**: high LR drives the model to extreme saturation, so the optimizer step IS updating parameters.
2. **The wrong parameters are getting all the gradient signal**: top-1 stays at 1/8 means the model output is input-INDEPENDENT. The 8 distinct inputs map to the same output byte.
3. **Therefore: input-dependent params (EmbeddingLayer weights + everything downstream that conditions on them) are NOT receiving useful gradients; only input-independent params (probably just the final Dense head's bias) are being updated.**

### Where to look next

The most likely culprit is the gradient flow through `Engine.TensorEmbeddingLookup` and/or the param-buffer-aliased path used by `TrainWithTape`. The EmbeddingLayer comments cite Tensors#255 and Tensors#257 as the fixes that made the tape track `dL/dE` correctly, but something downstream is dropping the input→embedding→attention→head gradient chain — leaving only the bias-update path live.

Specific next steps a debugger should take:
1. Hook `TapeTrainingStep.CollectParameters` and log the L2-norm of the gradient for every parameter tensor after one Train step. The hypothesis: embedding-weight gradient norm == 0 (or near-zero), Dense-bias gradient norm > 0.
2. If embedding gradient is zero: the issue is in `Engine.TensorEmbeddingLookup`'s backward registration (Tensors#255) — possibly broken by a recent Tensors version bump.
3. If embedding gradient is non-zero but the optimizer.Step doesn't apply it: `TryTrainWithFusedOptimizer` or the buffer-aliasing path may be skipping the embedding tensor (per the comment in NeuralNetworkBase.cs line 5046-5051 about "network-level raw tensors" being excluded from the buffer).

I have to stop here — going deeper into the Tensors-side tape internals + AutoTrainingCompiler is outside my session budget. Pushed branch + diagnostic test for whoever picks this up. Happy to continue if there's a specific direction to dig.

cc @ooples

**Comment by ooples** [comment]
## Bug ISOLATED — fused-compiled training kernel is the regression

Built production-grade training diagnostics (`TrainingDiagnosticsConfig`) mirroring the existing `GpuDiagnosticsConfig` three-knob pattern (env var + static + sink). Added an event-emission hook in `NeuralNetworkBase.TrainWithTape` and `TryTrainWithFusedOptimizer`, plus a `ForceEagerPath` knob.

### Smoking gun

With `TrainingDiagnosticsConfig.PerStepEnabled = true` on the existing `Transformer_ByteLM_DefaultPooling_TrainsBelowLnV` test config:

```
Captured 0 GradientNormEvent records, 0 TrainingLossEvent records.
```

`TrainWithTape` is never called. The Transformer training engages `TryTrainWithFusedOptimizer` and runs the entire forward + backward + optimizer step inside the compiled-plan kernel (`CompiledTapeTrainingStep.TryStepWithFusedOptimizer` → `plan.Step()`). The eager tape-walk diagnostic hook is below the fused-path branch and never fires.

### Confirming: fused = broken, eager = correct

New diagnostic test `Diagnostic_EagerPath_OnlyConvergeCheck` sets `TrainingDiagnosticsConfig.ForceEagerPath = true` (skips fused path at the top of `TryTrainWithFusedOptimizer`) and runs the exact same 500-epoch byte-LM convergence task:

```
EAGER-only path: avgNll=0.2023  ln(V)=2.0794  ratio=0.097  top-1=7/8=87.5%
  EAGER PATH CONVERGES — fused path is the regression.
```

vs. the fused-path run from the existing test which produced avgNll=2.5 (worse than uniform) and top-1=1/8 = 12.5% (input-independent output).

### Where the bug lives

The actual compiled kernel is in **AiDotNet.Tensors** — `plan.Step()` invokes the JIT-compiled forward + backward + fused-optimizer kernel. `CompiledTapeTrainingStep.cs` (this repo, line 308-336) only configures and replays the plan; the gradient flow and Adam state update happen entirely inside the Tensors-side kernel. Specifically:

- `CompiledModelCache<T>.GetOrCompileTraining` — caches the compiled plan keyed by input + target shape
- `plan.ConfigureOptimizer(...)` — attaches Adam m/v buffers to the plan
- `plan.Step()` — runs the entire training step in the compiled kernel

One of those (most likely the backward kernel for `TensorEmbeddingLookup` or the way Adam's m/v is wired into the EmbeddingLayer's weight tensor inside the plan) drops the embedding-weight gradient on the floor. The eager path uses `GradientTape.ComputeGradients` which DOES correctly populate `dL/dE` (per the comment in `EmbeddingLayer.cs` line 593-597 citing Tensors#255 / #257), so the divergence is specifically in the fused-replay path.

### Workaround shipping in this PR

```csharp
AiDotNet.Configuration.TrainingDiagnosticsConfig.ForceEagerPath = true;
```

Forces every Transformer Train step onto the eager tape-walk path. Same correctness as pre-fused-optimizer training, no compile-once fast path. Documented in the `TrainingDiagnosticsConfig` XML doc as the official workaround for #1328 until the Tensors kernel is fixed.

### PR

Branch [fix/1328-transformer-facade-training-degeneracy](https://github.com/ooples/AiDotNet/tree/fix/1328-transformer-facade-training-degeneracy) (latest commit 5f71fb2a2). Will open as PR shortly. Adds:

- `src/Configuration/TrainingDiagnosticLevel.cs`
- `src/Configuration/TrainingDiagnosticEvent.cs` (typed events using existing `LayerCategory` enum, not strings)
- `src/Configuration/TrainingDiagnosticSink.cs`
- `src/Configuration/TrainingDiagnosticsConfig.cs`
- Hooks in `NeuralNetworks/NeuralNetworkBase.cs` (TrainWithTape + TryTrainWithFusedOptimizer)
- 4 diagnostic tests in `tests/.../TransformerByteLMDiagnostic1328Tests.cs`

### Recommended next step on the Tensors side

Add a regression test in AiDotNet.Tensors that:
1. Compiles a tiny EmbeddingLayer → MultiHeadAttention → Dense → Softmax training plan
2. Runs one fused Step
3. Reads back the embedding weight gradient
4. Asserts it's non-zero for distinct one-hot inputs

That test currently passes in eager mode and fails in fused mode — pinning down which op in the compiled backward graph zeroes the embedding gradient should be quick once the regression test is in place.

cc @ooples

**Comment by ooples** [comment]
Upstream fix shipped: **AiDotNet.Tensors PR #354** (https://github.com/ooples/AiDotNet.Tensors/pull/354).

One-line fix in `CpuEngine.TensorEmbeddingLookup`: add the `GraphMode.IsActive` branch every other differentiable op uses, recording a Custom-type LazyNode with the existing `TensorEmbeddingLookupBackward` handler. The lazy graph now sees the op, so the compiled training plan produces a backward step for it, so dL/dE flows correctly through `plan.Step()`.

3 regression tests in `tests/.../Compilation/TensorEmbeddingLookupCompiledPlanTests.cs`:
- LazyNode is recorded under GraphMode
- Compiled plan produces non-zero gradient on selected rows (and zero on unselected)
- Distinct index batches produce distinct gradient patterns (rules out always-zero AND always-same regressions)

All 17 existing tape tests still pass — eager path untouched.

Once #354 merges + Tensors nuget publishes, consumers can drop the `TrainingDiagnosticsConfig.ForceEagerPath = true` workaround from PR #1330 and run the fused-compiled fast path correctly.
